lib/modelinstance.js

```diff
@@ -458,105 +465,118 @@ function _tryRemoveRefs(bucket, keys, callback) {
-      var removedRefKeys = [];
-      for (var i = 0; i < newRefKeys.length; ++i) {
-        if (oldRefKeys.indexOf(newRefKeys[i]) === -1) {
-          addedRefKeys.push(newRefKeys[i]);
-        }
-      }
-      for (var j = 0; j < oldRefKeys.length; ++j) {
-        if (newRefKeys.indexOf(oldRefKeys[j]) === -1) {
-          removedRefKeys.push(oldRefKeys[j]);
-        }
-      }
-
-      _tryAddRefs($.schema.store, addedRefKeys, $.key, function (err) {
+      $.schema.validate(self, function (err) {
         if (err) {
-          callback(err);
+          reject(err);
           return;
         }
 
-        $.schema.store.store(newKey, newData, $.cas, function (err, cas) {
+        var newKey = _modelKey(self);
+        var newData = self.toCoo();
+        var newRefKeys = $.schema.refKeys(self);
+
+        var oldRefKeys = $.refKeys;
+        var addedRefKeys = [];
+        var removedRefKeys = [];
+        for (var i = 0; i < newRefKeys.length; ++i) {
+          if (oldRefKeys.indexOf(newRefKeys[i]) === -1) {
+            addedRefKeys.push(newRefKeys[i]);
+          }
+        }
+        for (var j = 0; j < oldRefKeys.length; ++j) {
+          if (newRefKeys.indexOf(oldRefKeys[j]) === -1) {
+            removedRefKeys.push(oldRefKeys[j]);
+          }
+        }
+
+        _tryAddRefs($.schema.store, addedRefKeys, $.key, function (err) {
           if (err) {
-            callback(err);
+            reject(err);
             return;
           }
 
-          $.cas = cas;
-          $.refKeys = newRefKeys;
-
-          _tryRemoveRefs($.schema.store, removedRefKeys, function (err) {
+          $.schema.store.store(newKey, newData, $.cas, function (err, cas) {
             if (err) {
-              callback(err);
+              reject(err);
               return;
             }
 
-            $.schema.execPostHandlers('save', self, callback);
+            $.cas = cas;
+            $.refKeys = newRefKeys;
+
+            _tryRemoveRefs($.schema.store, removedRefKeys, function (err) {
+              if (err) {
+                reject(err);
+                return;
+              }
+
+              $.schema.execPostHandlers('save', self, function (err, result) {
+                if (err) {
+                  reject(err);
+                  return;
+                }
+                resolve(result);
+              });
+            });
           });
         });
       });
     });
-  });
+  }).asCallback(callback);
 };
 
 /**
  * This is a helper method which allows us to load an array of
  *   model instances all at once.
  *
  * @param {ModelInstance[]} items
- * @param {Function} callback
+ * @param {Function=} callback
+ * @returns Promise
  */
 ModelInstance.loadAll = function (items, callback) {
-  if (!Array.isArray(items)) {
-    throw new Error('Must pass an array to loadAll');
-  }
+  return new Promise(function (resolve, reject) {
+    if (!Array.isArray(items)) {
+      reject(new Error('Must pass an array to loadAll'));
```

Comment:
warning -- missing return.   Code falls through here.  See above comments on the "return reject" idiom.  This is a very easy mistake to make, and the PR contains many, many such clauses, so make sure to do a search through the code to ensure all are caught.

If code falls through, in some cases you'll reject the promise then later resolve it.  This will be buried deep in some handler, user of ottoman will get bluebird warnings, or strange behavior, or both, which may be quite difficult to find.